src/v/compression/lz4_decompression_buffers.cc

```diff
@@ -74,4 +75,134 @@ size_t lz4_decompression_buffers::min_alloc_threshold() const {
+        return bufs.input_buffer.get();
+    case input_buffer_allocated:
+        bufs.state |= output_buffer_allocated;
+        st->allocated();
+        return bufs.output_buffer.get();
+    case both_buffers_allocated:
+    case output_buffer_allocated:
+        vassert(
+          false, "invalid allocation request when both buffers allocated");
+    }
+}
+
+// During a decompression operation this function is called via the LZ4F_free
+// wrapper. The function is typically called in the following sequence:
+// 1. When freeing the decompression context:
+//    a. free the tmp out buffer
+//    b. free the tmp in buffer
+//    c. free the decompression context
+// 2. When initializing the decompression context, this function will be called
+// on the two buffer addresses.
+// In all cases we either pass the address straight through to `free()` or if
+// the address is managed, we update the state. The state update ensures that
+// the next decompression operation starts with the correct state (no buffers
+// allocated)
+void free_lz4_obj(void* state, void* address) {
+    auto* st = static_cast<compression::lz4_decompression_buffers*>(state);
+
+    // decompression buffers have been processed. The free request is for the
+    // decompression c-struct. Fall back to free to release memory since we do
+    // not manage it.
+    if (!st->has_reserved_buffers()) {
+        st->pass_through_deallocated();
+        free(address);
+        return;
+    }
+
+    auto& bufs = st->buffers();
+
+    // If the address being freed does not match one of the static addresses we
+    // manage, fall back to free. This can happen because:
+    //
+    // 1. LZ4 frees memory before performing each allocation, resulting in
+    // interspersed calls to free/malloc where the freed address was not
+    // allocated from our pool.
+    //
+    // 2. The allocation was not done via this allocator, eg for blocks
+    // small enough that they should not be managed by custom allocator.
+    //
+    // In both cases these memory addresses will not match our managed buffers.
+    if (!bufs.is_managed_address(address)) {
+        st->pass_through_deallocated();
+        free(address);
+        return;
+    }
+
+    // Buffers are released by lz4 in the order: input buffer, output buffer,
+    // decompression ctx. The first two calls update the state here. The third
+    // call is passed through to free because we do not allocate memory for the
+    // decompression ctx.
+    switch (bufs.state) {
+        using enum compression::lz4_decompression_buffers::alloc_ctx::
+          allocation_state;
+    case no_buffers_allocated:
+    case output_buffer_allocated:
+        vassert(
+          false, "unexpected buffer state {} during deallocation", bufs.state);
+    case input_buffer_allocated:
+        st->deallocated();
+        bufs.state &= (~input_buffer_allocated);
+        return;
+    case both_buffers_allocated:
+        st->deallocated();
```

Comment:
Just curious, does lz4 always call `free_lz4_obj`, even the decompression failed? And do the buffers need to be cleared at all?

Reply:
lz4 does not guarantee to free memory, the user has to free it, it has the function `LZ4F_freeDecompressionContext` to do this. In our case it is guaranteed because we wrap the decompression context in a unique ptr with a custom deleter, this deleter then calls `LZ4F_freeDecompressionContext` for us when the unique ptr is destroyed, so `free_lz4_obj` will always be called via `LZ4F_freeDecompressionContext -> LZ4_Free -> free_lz4_obj`

Reply:
This is the relevant type definition https://github.com/redpanda-data/redpanda/blob/dev/src/v/compression/internal/lz4_frame_compressor.cc#L53-L59